File: taurus-protect-sdk-java/client/src/test/java/com/taurushq/sdk/protect/client/helper/VerificationBehaviourVectorsTest.java
package com.taurushq.sdk.protect.client.helper;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.taurushq.sdk.protect.client.model.WhitelistException;
import com.taurushq.sdk.protect.client.model.WhitelistSignature;
import com.taurushq.sdk.protect.client.model.rulescontainer.AddressWhitelistingRules;
import com.taurushq.sdk.protect.client.model.rulescontainer.ContractAddressWhitelistingRules;
import com.taurushq.sdk.protect.client.model.rulescontainer.DecodedRulesContainer;
import com.taurushq.sdk.protect.openapi.auth.CryptoTPV1;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * Cross-SDK behaviour vectors for the verification primitives.
 *
 * <p>These are the invariants that had drifted apart before: which (blockchain,
 * network) pair selects the governance rules, and whether a hash is covered by a
 * signature. Both are pure input to outcome mappings, so all four SDKs assert them
 * from one file rather than from four hand-maintained copies — the arrangement that
 * let this SDK's hash-coverage go non-constant-time while its peers did not.
 *
 * <p>To add a case: append to the shared file, bump the matching count there, and
 * consume it in all four suites.
 */
class VerificationBehaviourVectorsTest {

    private static JsonObject vectors;

    @BeforeAll
    static void loadVectors() throws IOException {
        String[] candidates = {
                "../../scripts/resources/verification-behaviour-vectors.json", // from client/
                "../scripts/resources/verification-behaviour-vectors.json",    // from sdk-java/
                "scripts/resources/verification-behaviour-vectors.json",       // from repo root
        };
        for (String candidate : candidates) {
            Path p = Paths.get(candidate).toAbsolutePath().normalize();
            if (Files.exists(p)) {
                String json = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
                vectors = JsonParser.parseString(json).getAsJsonObject();
                assertCounts();
                return;
            }
        }
        throw new IOException("Cannot find verification-behaviour-vectors.json. Working directory: "
                + System.getProperty("user.dir"));
    }

    /**
     * Counts are asserted so a case added to the shared file without being consumed
     * here fails loudly rather than being silently ignored by this SDK.
     */
    private static void assertCounts() {
        JsonObject counts = vectors.getAsJsonObject("counts");
        for (String key : counts.keySet()) {
            assertEquals(counts.get(key).getAsInt(), vectors.getAsJsonArray(key).size(),
                    key + ": vector count disagrees with the count the file declares");
        }
    }

    @Test
    void resolvesTheGovernanceRuleKeyForEveryVector() {
        for (JsonElement element : vectors.getAsJsonArray("rule_key")) {
            JsonObject c = element.getAsJsonObject();
            String description = c.get("description").getAsString();
            String payload = c.get("payload_as_string").getAsString();
            String dtoBlockchain = c.get("dto_blockchain").getAsString();
            String dtoNetwork = c.get("dto_network").getAsString();

            if ("error".equals(c.get("expect").getAsString())) {
                assertThrows(WhitelistException.class,
                        () -> WhitelistHashHelper.resolveRuleKey(payload, dtoBlockchain, dtoNetwork),
                        description);
                continue;
            }

            String[] resolved;
            try {
                resolved = WhitelistHashHelper.resolveRuleKey(payload, dtoBlockchain, dtoNetwork);
            } catch (WhitelistException e) {
                throw new AssertionError(description + ": unexpected failure", e);
            }
            assertEquals(c.get("blockchain").getAsString(), resolved[0], description);
            assertEquals(c.get("network").getAsString(), resolved[1], description);
        }
    }

    @Test
    void agreesOnHashCoverageForEveryVector() {
        for (JsonElement element : vectors.getAsJsonArray("hash_coverage")) {
            JsonObject c = element.getAsJsonObject();
            String description = c.get("description").getAsString();

            List<WhitelistSignature> signatures = new ArrayList<>();
            for (JsonElement sigElement : c.getAsJsonArray("signatures")) {
                WhitelistSignature signature = new WhitelistSignature();
                for (JsonElement hash : sigElement.getAsJsonArray()) {
                    signature.getHashes().add(hash.getAsString());
                }
                signatures.add(signature);
            }

            assertEquals(c.get("expect").getAsBoolean(),
                    SignatureVerifier.verifyHashCoverage(c.get("hash").getAsString(), signatures),
                    description);
        }
    }

    @Test
    void agreesOnSingleSignatureHashContainmentForEveryVector() {
        for (JsonElement element : vectors.getAsJsonArray("contains_hash")) {
            JsonObject c = element.getAsJsonObject();
            String description = c.get("description").getAsString();

            List<String> hashes = new ArrayList<>();
            for (JsonElement hash : c.getAsJsonArray("hashes")) {
                hashes.add(hash.getAsString());
            }

            assertEquals(c.get("expect").getAsBoolean(),
                    SignatureVerifier.containsHash(hashes, c.get("hash").getAsString()),
                    description);
        }
    }

    // ------------------------------------------------------------------------------
    // legacy_hash and rule_tier_candidates
    // ------------------------------------------------------------------------------
    //
    // Until 2026-09-10 both sections were consumed by the GO suite alone, even though the
    // file's `counts` block declares them and the loop above asserts every count.
    // Asserting a section's LENGTH proves the file is well formed; it does not prove the
    // behaviour is checked. Python shipped the single-tier lookup `rule_tier_candidates`
    // exists to forbid while three SDKs had the fix, and no gate went red.

    @Test
    @DisplayName("step 6 parses the payload the signature COVERED, for every vector")
    void parsesThePayloadTheSignatureCovered() {
        // Asserts the PARSED PAYLOAD, not a hash: the legacy-strip injection moves no hash
        // at all, so crypto-test-vectors.json is structurally blind to it.
        for (JsonElement e : vectors.getAsJsonArray("legacy_hash")) {
            JsonObject c = e.getAsJsonObject();
            String description = c.get("description").getAsString();
            String coveredHash =
                    CryptoTPV1.calculateHexHash(c.get("signed_payload").getAsString());
            String delivered = c.get("delivered_payload").getAsString();

            String matchedPayload = null;
            if (coveredHash.equals(CryptoTPV1.calculateHexHash(delivered))) {
                matchedPayload = delivered;
            } else {
                for (LegacyPayloadVariant variant
                        : WhitelistHashHelper.computeLegacyPayloadVariants(delivered)) {
                    if (coveredHash.equals(variant.getHash())) {
                        matchedPayload = variant.getPayload();
                        break;
                    }
                }
            }

            if ("no_match".equals(c.get("expect").getAsString())) {
                assertNull(matchedPayload,
                        description + ": expected no variant to be covered, but one matched");
                continue;
            }
            assertNotNull(matchedPayload, description + ": expected a covered variant");
            assertEquals(c.get("expect_matched_payload").getAsString(), matchedPayload,
                    description);
        }
    }

    @Test
    @DisplayName("every reachable rule tier is returned, for every vector")
    void returnsEveryReachableRuleTier() {
        // When the signed payload omits `network` there is no authenticated way to learn
        // whether that was legitimate, so every tier the unsigned DTO value could have
        // selected must be enforced. Too few lets the server pick the quorum; too many
        // rejects rows governance would accept.
        for (JsonElement e : vectors.getAsJsonArray("rule_tier_candidates")) {
            JsonObject c = e.getAsJsonObject();
            String description = c.get("description").getAsString();
            String blockchain = c.get("blockchain").getAsString();

            List<AddressWhitelistingRules> addressRules = new ArrayList<>();
            List<ContractAddressWhitelistingRules> contractRules = new ArrayList<>();
            for (JsonElement r : c.getAsJsonArray("rules")) {
                JsonObject rule = r.getAsJsonObject();
                AddressWhitelistingRules a = new AddressWhitelistingRules();
                a.setCurrency(rule.get("blockchain").getAsString());
                a.setNetwork(rule.get("network").getAsString());
                addressRules.add(a);

                ContractAddressWhitelistingRules k = new ContractAddressWhitelistingRules();
                k.setBlockchain(rule.get("blockchain").getAsString());
                k.setNetwork(rule.get("network").getAsString());
                contractRules.add(k);
            }
            DecodedRulesContainer container = new DecodedRulesContainer();
            container.setAddressWhitelistingRules(addressRules);
            container.setContractAddressWhitelistingRules(contractRules);

            List<String> expected = new ArrayList<>();
            for (JsonElement x : c.getAsJsonArray("expect_candidates")) {
                JsonObject want = x.getAsJsonObject();
                expected.add(want.get("blockchain").getAsString() + "|"
                        + want.get("network").getAsString());
            }

            List<String> gotAddress = new ArrayList<>();
            for (AddressWhitelistingRules r
                    : container.findAddressWhitelistingRuleCandidates(blockchain)) {
                gotAddress.add(nullToEmpty(r.getCurrency()) + "|" + nullToEmpty(r.getNetwork()));
            }
            assertEquals(expected, gotAddress, description);

            // The asset peer, from the same vectors: the two families name the chain field
            // differently (currency vs blockchain), so a walk reading one name treats every
            // contract rule as a wildcard global default — fail-OPEN, the broadest tier.
            List<String> gotContract = new ArrayList<>();
            for (ContractAddressWhitelistingRules r
                    : container.findContractAddressWhitelistingRuleCandidates(blockchain)) {
                gotContract.add(nullToEmpty(r.getBlockchain()) + "|" + nullToEmpty(r.getNetwork()));
            }
            assertEquals(expected, gotContract, description + " (contract family)");
        }
    }

    private static String nullToEmpty(final String value) {
        return value == null ? "" : value;
    }
}

File: taurus-protect-sdk-go/pkg/protect/helper/verification_behaviour_vectors_test.go
package helper

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/crypto"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// Cross-SDK behaviour vectors for the verification primitives.
//
// These are the invariants that had drifted apart before: which (blockchain,
// network) pair selects the governance rules, and whether a hash is covered by a
// signature. Both are pure input -> outcome mappings, so all four SDKs assert them
// from one file rather than from four hand-maintained copies — the arrangement that
// let Java's hash-coverage go non-constant-time while its peers did not.
//
// To add a case: append to the shared file, bump the matching count there, and
// consume it in all four suites.
const behaviourVectorsRelPath = "../../../../scripts/resources/verification-behaviour-vectors.json"

type behaviourVectors struct {
	Counts struct {
		RuleKey            int `json:"rule_key"`
		HashCoverage       int `json:"hash_coverage"`
		ContainsHash       int `json:"contains_hash"`
		LegacyHash         int `json:"legacy_hash"`
		RuleTierCandidates int `json:"rule_tier_candidates"`
	} `json:"counts"`
	LegacyHash []struct {
		Description          string `json:"description"`
		SignedPayload        string `json:"signed_payload"`
		DeliveredPayload     string `json:"delivered_payload"`
		Expect               string `json:"expect"`
		ExpectMatchedPayload string `json:"expect_matched_payload"`
	} `json:"legacy_hash"`
	RuleTierCandidates []struct {
		Description string `json:"description"`
		Rules       []struct {
			Blockchain string `json:"blockchain"`
			Network    string `json:"network"`
		} `json:"rules"`
		Blockchain       string `json:"blockchain"`
		ExpectCandidates []struct {
			Blockchain string `json:"blockchain"`
			Network    string `json:"network"`
		} `json:"expect_candidates"`
	} `json:"rule_tier_candidates"`
	RuleKey []struct {
		Description     string `json:"description"`
		PayloadAsString string `json:"payload_as_string"`
		DTOBlockchain   string `json:"dto_blockchain"`
		DTONetwork      string `json:"dto_network"`
		Expect          string `json:"expect"`
		Blockchain      string `json:"blockchain"`
		Network         string `json:"network"`
	} `json:"rule_key"`
	HashCoverage []struct {
		Description string     `json:"description"`
		Signatures  [][]string `json:"signatures"`
		Hash        string     `json:"hash"`
		Expect      bool       `json:"expect"`
	} `json:"hash_coverage"`
	ContainsHash []struct {
		Description string   `json:"description"`
		Hashes      []string `json:"hashes"`
		Hash        string   `json:"hash"`
		Expect      bool     `json:"expect"`
	} `json:"contains_hash"`
}

func loadBehaviourVectors(t *testing.T) behaviourVectors {
	t.Helper()

	path := filepath.Clean(behaviourVectorsRelPath)
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("cannot read shared verification behaviour vectors %s: %v", path, err)
	}

	var v behaviourVectors
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("cannot parse %s: %v", path, err)
	}

	// Counts are asserted so a case added to the shared file without being consumed
	// here fails loudly rather than being silently ignored by this SDK.
	if len(v.RuleKey) != v.Counts.RuleKey {
		t.Fatalf("rule_key: got %d vectors, file declares %d", len(v.RuleKey), v.Counts.RuleKey)
	}
	if len(v.HashCoverage) != v.Counts.HashCoverage {
		t.Fatalf("hash_coverage: got %d vectors, file declares %d", len(v.HashCoverage), v.Counts.HashCoverage)
	}
	if len(v.ContainsHash) != v.Counts.ContainsHash {
		t.Fatalf("contains_hash: got %d vectors, file declares %d", len(v.ContainsHash), v.Counts.ContainsHash)
	}
	if len(v.LegacyHash) != v.Counts.LegacyHash {
		t.Fatalf("legacy_hash: got %d vectors, file declares %d", len(v.LegacyHash), v.Counts.LegacyHash)
	}
	if len(v.RuleTierCandidates) != v.Counts.RuleTierCandidates {
		t.Fatalf("rule_tier_candidates: got %d vectors, file declares %d",
			len(v.RuleTierCandidates), v.Counts.RuleTierCandidates)
	}
	return v
}

// TestVerificationBehaviourVectors_LegacyHash pins WHICH PAYLOAD step 6 must parse.
//
// The vector gives the payload a signer covered and the payload the server delivered. The SDK
// must decide, from the delivered payload alone, which variant a signature over the signed
// payload covers — and hand back THAT payload. Asserting the hash alone (which is all
// crypto-test-vectors.json does) cannot catch the defect, because the attack does not move any
// hash: it changes which string the parse reads.
func TestVerificationBehaviourVectors_LegacyHash(t *testing.T) {
	for _, tc := range loadBehaviourVectors(t).LegacyHash {
		t.Run(tc.Description, func(t *testing.T) {
			coveredHash := crypto.CalculateHexHash(tc.SignedPayload)

			matchedPayload := ""
			matched := false
			if crypto.CalculateHexHash(tc.DeliveredPayload) == coveredHash {
				matchedPayload, matched = tc.DeliveredPayload, true
			} else {
				for _, variant := range ComputeLegacyPayloadVariants(tc.DeliveredPayload) {
					if variant.Hash == coveredHash {
						matchedPayload, matched = variant.Payload, true
						break
					}
				}
			}

			if tc.Expect == "no_match" {
				if matched {
					t.Fatalf("expected no variant to be covered, but %q matched", matchedPayload)
				}
				return
			}
			if !matched {
				t.Fatal("expected a covered variant, found none")
			}
			if matchedPayload != tc.ExpectMatchedPayload {
				t.Errorf("matched payload:\n got  %q\n want %q", matchedPayload, tc.ExpectMatchedPayload)
			}
		})
	}
}

// TestVerificationBehaviourVectors_RuleTierCandidates pins the set of rules that must be enforced
// when the signed payload omits `network`. Too few leaves the server free to pick the quorum; too
// many rejects rows governance would accept.
func TestVerificationBehaviourVectors_RuleTierCandidates(t *testing.T) {
	for _, tc := range loadBehaviourVectors(t).RuleTierCandidates {
		t.Run(tc.Description, func(t *testing.T) {
			container := &model.DecodedRulesContainer{}
			for _, r := range tc.Rules {
				container.AddressWhitelistingRules = append(container.AddressWhitelistingRules,
					&model.AddressWhitelistingRules{Currency: r.Blockchain, Network: r.Network})
			}

			got := container.FindAddressWhitelistingRuleCandidates(tc.Blockchain)
			if len(got) != len(tc.ExpectCandidates) {
				t.Fatalf("got %d candidates, want %d", len(got), len(tc.ExpectCandidates))
			}
			for i, want := range tc.ExpectCandidates {
				if got[i].Currency != want.Blockchain || got[i].Network != want.Network {
					t.Errorf("candidate %d = %q/%q, want %q/%q",
						i, got[i].Currency, got[i].Network, want.Blockchain, want.Network)
				}
			}
		})
	}
}

func TestVerificationBehaviourVectors_RuleKey(t *testing.T) {
	for _, tc := range loadBehaviourVectors(t).RuleKey {
		t.Run(tc.Description, func(t *testing.T) {
			blockchain, network, err := ResolveRuleKey(tc.PayloadAsString, tc.DTOBlockchain, tc.DTONetwork)

			if tc.Expect == "error" {
				if err == nil {
					t.Fatalf("expected an error, got %q/%q", blockchain, network)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if blockchain != tc.Blockchain || network != tc.Network {
				t.Errorf("got %q/%q, want %q/%q", blockchain, network, tc.Blockchain, tc.Network)
			}
		})
	}
}

func TestVerificationBehaviourVectors_HashCoverage(t *testing.T) {
	for _, tc := range loadBehaviourVectors(t).HashCoverage {
		t.Run(tc.Description, func(t *testing.T) {
			signatures := make([]model.WhitelistSignature, 0, len(tc.Signatures))
			for _, hashes := range tc.Signatures {
				signatures = append(signatures, model.WhitelistSignature{Hashes: hashes})
			}

			if got := VerifyHashCoverage(tc.Hash, signatures); got != tc.Expect {
				t.Errorf("VerifyHashCoverage(%q) = %v, want %v", tc.Hash, got, tc.Expect)
			}
		})
	}
}

func TestVerificationBehaviourVectors_ContainsHash(t *testing.T) {
	for _, tc := range loadBehaviourVectors(t).ContainsHash {
		t.Run(tc.Description, func(t *testing.T) {
			if got := containsHash(tc.Hashes, tc.Hash); got != tc.Expect {
				t.Errorf("containsHash(%v, %q) = %v, want %v", tc.Hashes, tc.Hash, got, tc.Expect)
			}
		})
	}
}
